src/eval.c

```diff
@@ -5394,6 +5465,779 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    char_u*		result;
+
+    if ((quoted[0] != '\'') && (quoted[0] != '"'))
+    {
+	return NULL;
+    }
+    result = alloc_terminated(result_length);
+
+    memcpy(result, quoted + 1, result_length);
+    return result;
+}
+
+/*
+ * This result MUST BE `free`d later!
+ */
+    static char_u*
+duplicate_single_quote(char_u x)
+{
+    char_u*	xs;
+
+    if (x == '\'')
+    {
+	return vim_strsave((char_u*) "''");
+    }
+
+    xs = alloc_terminated(1);
+    xs[0] = x;
+    return xs;
+}
+
+/*
+ * This result MUST BE `free`d later!
+ */
+    static char_u*
+dont_fix(char_u x)
+{
+    char_u*	xs = alloc_terminated(1);
+    xs[0] = x;
+    return xs;
+}
+
+    static int
+is_closing_single_quote(char_u *x)
+{
+    const char_u	DELIMITTERS[] = {NUL, ',', ')', ' '};
+    const size_t	DELIMITTERS_SIZE = 4;
+    size_t		i;
+
+    if (x[0] != '\'')
+    {
+	return FALSE;
+    }
+
+    for (i = 0; i < DELIMITTERS_SIZE; ++i)
+    {
+	if (x[1] == DELIMITTERS[i])
+	{
+	    return TRUE;
+	}
+    }
+
+    return FALSE;
+}
+
+/*
+ * NOTICE: x[-1] must be readable.
+ */
+    static int
+is_closing_double_quote(char_u *x)
+{
+    return (x[0] == '"') && (x[-1] != '\\');
+}
+
+/*
+ * Allocate a variable for a $'string ${var} $x' constant.
+ * Return OK or FAIL.
+ *
+ * enclose() encloses a content by '' (or "").
+ * fix_quote() fixes quotes ' to ''. this fixes a problem that string() stringifies {"x": 10} to {'x': 10}, ["a", "b"] to ['a', 'b'].
+ * is_closing_quote() checks a char is ' (or ").
+ */
+    static int
+get_template_string_tv(
+	char_u* (*enclose)(char_u*),
+	char_u* (*disclose)(char_u*),
+	char_u* (*fix_quote)(char_u),
+	int (*is_closing_quote)(char_u*),
+	char_u **arg,
+	typval_T *rettv,
+	int evaluate)
+{
+    char_u*	string;
+    char_u*	origin_string;
+    char_u*	template;
+
+    // NOTE: At here, "${}" of "${" "expr" "}", that called 'embedding', and "expr" called 'embedded'.
+
+    template = read_template_string(is_closing_quote, *arg);
+    if (template == NULL)
+    {
+	return FAIL;
+    }
+
+    // Parse a template string into a plain string
+    string = parse_template(enclose, disclose, fix_quote, is_closing_quote, template, evaluate);
+    if (string == NULL)
+    {
+	return FAIL;
+    }
+
+    // Eval the result (non template) string
+    origin_string = string;
+    eval1(&string, rettv, evaluate);
+    free(origin_string);
+
+    // Consume
+    *arg += STRLEN(template);
+    free(template);
+
+    return OK;
+}
+
+/*
+ * Get a string between a opening "$'" and a closing "'" (or "$\"" and "\""),
+ * and check a taken string is not missing a closing "'".
+ *
+ * This result MUST BE `free`d later! (if it is not NULL.)
+ */
+    static char_u*
+read_template_string(int (*is_closing_quote)(char_u*), char_u *source)
+{
+    const size_t 	OPENING_QUOTE_LENGTH = 2;  // "$'"
+    char_u*		p;
+    char_u*		template;
+    size_t		template_length;
+
+    for (p = source + OPENING_QUOTE_LENGTH; !(is_closing_quote(p) || *p == NUL); MB_PTR_ADV(p));
```

Comment:
It might be better to move the last `;` to the next line (with indent) to emphasize that this is an empty loop.

Reply:
I'll do it :D